youtube_dl/extractor/hketv.py

```diff
@@ -0,0 +1,160 @@
+        'info_dict': {
+            'id': '2932360618',
+            'ext': 'mp4',
+            'title': '喜閱一生(共享閱讀樂)',
+            'description': '本節目輯錄了「閱讀滿Fun嘉年華」和「二○一八響應世界閱讀日――悅愛閱讀・愈讀愈愛」的活動花絮，並由學者、作家、演藝界人士等，分享培養子女閱讀興趣和習慣的方法，以及呼籲大家一同分享閱讀的樂趣。',
+            'upload_date': '20181024',
+            'duration': 900,
+            'subtitles': {
+                'en': [{
+                    'url': 'https://apps.hkedcity.net/media/mediaplayer/caption.php?f=74395&lang=en',
+                    'ext': 'srt',
+                }],
+                'zh-Hant': [{
+                    'url': 'https://apps.hkedcity.net/media/mediaplayer/caption.php?f=74395&lang=qmt',
+                    'ext': 'srt',
+                }],
+            }
+        },
+    }]
+
+    _CC_LANGS = {
+        '中文（繁體中文）': 'zh-Hant',
+        '中文（简体中文）': 'zh-Hans',
+        'English': 'en',
+        'Bahasa Indonesia': 'id',
+        '\u0939\u093f\u0928\u094d\u0926\u0940': 'hi',
+        '\u0928\u0947\u092a\u093e\u0932\u0940': 'ne',
+        'Tagalog': 'tl',
+        '\u0e44\u0e17\u0e22': 'th',
+        '\u0627\u0631\u062f\u0648': 'ur',
+    }
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+
+        file_id = self._html_search_regex(r'post_var\["file_id"\] = ([0-9]+);', webpage, 'file ID')
+        curr_url = self._html_search_regex(r'post_var\["curr_url"\] = "(.+?)";', webpage, 'curr URL')
+        data = {
+            'action': 'get_info',
+            'curr_url': curr_url,
+            'file_id': file_id,
+            'video_url': file_id,
+        }
+        _APPS_BASE_URL = 'https://apps.hkedcity.net'
+        handler_url = urljoin(_APPS_BASE_URL, '/media/play/handler.php')
+
+        response = self._download_json(
+            handler_url, video_id,
+            data=urlencode_postdata(data),
+            headers=merge_dicts({'Content-Type': 'application/x-www-form-urlencoded'},
+                                self.geo_verification_headers()))
+
+        result = try_get(response, lambda x: x['result'], dict)
+
+        formats = []
+        subtitles = {}
+
+        thumbnail_php = urljoin(_APPS_BASE_URL, result.get('image'))
+        thumbnail_urlh = self._downloader.urlopen(thumbnail_php)
+        thumbnail = thumbnail_urlh.geturl()
+
+        if response.get('success') and response.get('access'):
+            width = str_to_int(result.get('width'))
+            height = str_to_int(result.get('height'))
+
+            playlist0 = try_get(result, lambda x: x['playlist'][0], dict)
+            fmts = try_get(playlist0, lambda x: x['sources'], list)
+            for fmt in fmts:
+                label = fmt.get('label')
+                if label == 'HD':
+                    h = 720
+                elif label == 'SD':
+                    h = 360
+                w = h * width // height
+                urlh = self._downloader.urlopen(urljoin(_APPS_BASE_URL, fmt.get('file')))
```

Comment:
`urljoin` replaced with `+`.

Or did you mean the use of `self._downloader.urlopen()`?  That was intentionally done so that the download link of the video could be downloaded without geo-restriction, i.e. without proxy, and without the PHPSESSID cookie.  (The geo-restriction is done within their PHP script only.)

Reply:
So, may I keep using self._downloader.urlopen() for the MPEG-4 video file?  :-)

Reply:
No, what for?

Reply:
So that no cookie is needed with the link given by `-g`.  The link given by `urljoin(_APPS_BASE_URL, fmt.get('file'))` looks something like this:

https://apps.hkedcity.net/media/mediaplayer/filename.php?t=85ee74dae3477d95b05cc3091f26ff5a&m=mp4&p=fc61b1acf92b52238ad5c616409759360b4d914f183691d79c03e27a335a1c7e99a0f5d12b7183530daad9edc79fb17d97c3b6c0d3c445d68b96c92fa88b3a2f&file=551cc229795bb70003e9962bda09899612a7ae10b183be5e84e136419d91c684c63c863a8e3237a753ed31921500c2d28aa4db43a8684e37caac000571948110&e=.mp4

which requires the correct cookies to work.  It gives a "HTTP/1.1 302 Found" redirecting to a final URL that looks like https://video1.hkedcity.net/streaming/b3848d0a3296b85d67d88b749642c29a/5c3738d1/channels/etv/201507/20150716151039_989507_720.mp4 which requires no cookie.  This final URL is easier for me to copy-and-paste to other programs or even to stream to other devices.

That said, I have decided to take your advice and not "fully resolve" to the final URL because the intermediate URL, with the appropriate cookies saved with `--cookies`, remains valid for many hours, whereas the fully resolved URL would be "410 Gone" within one or two hours.  Also, I realized `self._downloader.urlopen` was being run twice in the `for fmt in fmts:` loop, leading to quite a few seconds of delay.  (The HKETV server seems rather slow in responding to this request.)